src/libxtp/segment.cc

```diff
@@ -0,0 +1,542 @@
+/*
+ *            Copyright 2009-2018 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <vector>
+#include <string>
+
+#include <votca/xtp/polarsite.h>
+#include <votca/xtp/apolarsite.h>
+#include <votca/xtp/atom.h>
+#include <votca/xtp/fragment.h>
+#include <votca/xtp/segment.h>
+#include <votca/xtp/segmenttype.h>
+#include <votca/tools/vec.h>
+
+using namespace std;
+using namespace votca::tools;
+
+namespace votca {
+namespace xtp {
+
+/// Default constructor
+Segment::Segment(int id, string name)
+    : _id(id),
+      _name(name),
+      _has_e(false),
+      _has_h(false),
+      _has_s(false),
+      _has_t(false) {
+  _eMpoles.resize(5);
+}
+
+// This constructor creates a copy of the stencil segment, without
+// adding it to any containers further up in the hierarchy; i.e. the topology
+// and molecules will neither know about the existence of this segment, nor
+// be able to access it. Used for creating the ghost in PB corrected pairs.
+Segment::Segment(Segment *stencil)
+    : _id(stencil->getId()),
+      _name(stencil->getName() + "_ghost"),
+      _typ(stencil->getType()),
+      _top(NULL),
+      _mol(NULL),
+      _CoM(stencil->getPos()),
+      _has_e(false),
+      _has_h(false),
+      _has_s(false),
+      _has_t(false) {
+  _eMpoles.resize(5);
+
+  vector<Fragment *>::iterator fit;
+  for (fit = stencil->Fragments().begin(); fit < stencil->Fragments().end();
+       fit++) {
+
+    Fragment *newFrag = new Fragment(*fit);
+    this->AddFragment(newFrag);
+
+    vector<Atom *>::iterator ait;
+    for (ait = newFrag->Atoms().begin(); ait < newFrag->Atoms().end(); ait++) {
+      this->AddAtom(*ait);
+    }
+  }
+}
+
+Segment::~Segment() {
+
+  vector<Fragment *>::iterator fragit;
+  for (fragit = this->Fragments().begin(); fragit < this->Fragments().end();
+       fragit++) {
+    delete *fragit;
+  }
+  _fragments.clear();
+  _atoms.clear();
+
+  _eMpoles.clear();
+  _polarSites.clear();
+}
+
+void Segment::TranslateBy(const vec &shift) {
+
+  _CoM = _CoM + shift;
+
+  vector<Fragment *>::iterator fit;
+  for (fit = _fragments.begin(); fit < _fragments.end(); fit++) {
+
+    (*fit)->TranslateBy(shift);
+  }
+}
+
+void Segment::setHasState(bool yesno, int state) {
```

Comment:
I do not like this accessor way, can we discuss using maps instead, which have hole, electron, gs, singlet, triplet as keys or (e,h,n,s,t)

Reply:
I don't want to change any of the methods at this stage. I would prefer to wait until I get the dependence of xtp on these classes switched and working and then refactor, but yes let's discuss. What is this method doing? This is the one that is describing if the state is charged and or excited?

Reply:
yes, but actually we should aim for a better separation of state and topology, I still do not fully get what is what